src/Symfony/Component/Mime/Encoder/PunycodeAddressEncoder.php

```diff
@@ -0,0 +1,61 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Mime\Encoder;
+
+use Symfony\Component\Mime\Exception\AddressEncoderException;
+
+/**
+ * A punycode IDN email address encoder.
+ *
+ * This encoder does supports email addresses with non-ASCII characters in
+ * local-part (the substring before @).
+ *
+ * @author Saif Eddin Gmati <saif.gmati@symfony.com>
+ */
+final class PunycodeAddressEncoder implements AddressEncoderInterface
+{
+    private $encoder;
+
+    public function __construct(AddressEncoderInterface $encoder = null)
+    {
+        $this->encoder = $encoder ?? new IdnAddressEncoder();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function encodeString(string $address): string
+    {
+        $i = strrpos($address, '@');
+        if (false !== $i) {
+            $local = $this->encode(substr($address, 0, $i));
+            $domain = $this->encode(substr($address, $i + 1));
+
+            $address = $local.'@'.$domain;
+        }
+
+        return $this->encoder->encodeString($address);
+    }
+
+    private function encode(string $str): string
+    {
+        if (!\defined('INTL_IDNA_VARIANT_UTS46') && preg_match('/[\x80-\xFF]/', $str)) {
+            throw new AddressEncoderException(sprintf('Unsupported IDN "%s", try enabling the "intl" PHP extension or running "composer require symfony/polyfill-intl-idn".', $str));
+        }
+
+        if (false === $str = \defined('INTL_IDNA_VARIANT_UTS46') ? idn_to_ascii($str, IDNA_DEFAULT, INTL_IDNA_VARIANT_UTS46) : strtolower($str)) {
+            throw new AddressEncoderException(sprintf('Unsupported IDN "%s".', $str));
```

Comment:
wouldn't this use `false` to build the error message ?

Reply:
please also add tests covering this error case.

Reply:
> wouldn't this use false to build the error message ?

Lazy copy-paste from the http-client component :) needs a fix in there too i suppose.